### PR TITLE
Fix grammar verifcation

### PR DESCRIPTION
**Pull request by dns2utf8**
I wanted to improve the [grammar-section](https://doc.rust-lang.org/grammar.html#items-and-attributes) but I can not build it.

r? @steveklabnik @brson 

EDIT: I resolved some of the issues, removed the compiler output from the entry.

### CODE REVIEW

**Comment by dns2utf8** [comment]
I was able to fix the errors, yay :grin: The API of libsyntax was changed (eg 24e7491660845e52240afc36acf332b92547b0a1) but the veryfy.rs is not getting build.

Should we include the antlr4-complete.jar or just build the `verify.rs` with the compiler? So further changes to the API will triger an error.

File: src/grammar/README.md [hunk 1]
```diff
@@ -4,11 +4,11 @@ Uses [antlr4](http://www.antlr.org/) and a custom Rust tool to compare
 ASTs/token streams generated. You can use the `check-lexer` make target to
 run all of the available tests.
 
-To use manually:
+To use manually, assuming antler4 ist installed at `/usr/share/java/antlr-complete.jar`:
```
Comment (dbrgn) [inline src/grammar/README.md]: This should probably say "antlr", not "antler".

**Comment by dns2utf8** [comment]
I ran the `check-lexer` target and the result was this:

```bash
test result: failed. 4764 passed; 125 failed; 2 skipped
```
I am curious, what are the next steps to fix the faileds tests?

File: src/grammar/README.md (hunk 1 above)
Comment (dns2utf8) [inline src/grammar/README.md]: Ah, yes thank you :+1:

File: mk/grammar.mk [hunk 2]
```diff
@@ -37,7 +37,7 @@ $(BG):
 
 $(BG)RustLexer.class: $(BG) $(SG)RustLexer.g4
 	$(Q)$(CFG_ANTLR4) -o $(BG) $(SG)RustLexer.g4
-	$(Q)$(CFG_JAVAC) -d $(BG) $(BG)RustLexer.java
+	$(Q)$(CFG_JAVAC) -d $(BG) -classpath /usr/share/java/antlr-complete.jar $(BG)RustLexer.java
```
Comment (steveklabnik) [inline mk/grammar.mk]: hm, is this always going to be where this is? Feels system-specific

**Comment by steveklabnik** [comment]
@alexcrichton what do you think of this PR?

**Comment by alexcrichton** [comment]
It does seems system specific, perhaps that could be refactored? That being said I know very little about all of this, and we don't run it on the bots, so I don't mind merging basically whatever here.

**Comment by brson** [comment]
I'm cool with pushing these improvements through even with the hardcoded path. It's better than it was, and I'm glad somebody is looking at this stuff. Would be good to get this onto a bot https://github.com/rust-lang/rust/issues/28592

**Comment by dns2utf8** [comment]
The system specific path could be obmittet with an environement variable, so the `javac` can find the `antlr4-complete.jar` but this would require the user to set it manually or the configure script to search for it.

Another alternative would be to include the jar file and just ship it. Is this a good idea?

cc: @nikomatsakis

**Comment by dns2utf8** [comment]
I removed the auto build from the tidy checks for now. 

If this is ok, I can rebase it and continue from there.

File: mk/grammar.mk (hunk 2 above)
Comment (dns2utf8) [inline mk/grammar.mk]: It is the place in Arch Linux, but not on Mac OS X. I have not tested other distros.

**Comment by dns2utf8** [comment]
I add the target `check-build-lexer-verifier` to `make tidy`, so it will build the verifier with every build and catch future changes in the API.
This does not require `javac` or `antlr`.

I think this is a good state freeze/merge for now. I planed to add some documumentation about the process in the next weeks, but I am no sure whether it will be in a good shape before the 18. august or not.

**Comment by steveklabnik** [comment]
@bors: r+ rollup

:shipit: now, we can make it better eventually :smile:

File: mk/clean.mk [hunk 3]
```diff
@@ -47,6 +47,10 @@ clean-misc:
 	$(Q)rm -Rf dist/*
 	$(Q)rm -Rf doc
 
+clean-grammar:
+	@$(call E, cleaning grammar verification)
+	$(Q)cd src/grammar && rm -Rf verify *.class *.java *.tokens
```
Comment (jonathandturner) [inline mk/clean.mk]: This does not handle an out-of-tree build directory, which our bots use.

```
cleaning grammar verification
/bin/sh: 1: cd: can't cd to src/grammar
/buildslave/rust-buildbot/slave/auto-linux-64-nopt-t/build/mk/clean.mk:51: recipe for target 'clean-grammar' failed
make: *** [clean-grammar] Error 2
program finished with exit code 2
elapsedTime=0.958165
```

Comment (dns2utf8) [inline mk/clean.mk]: Interesting, looking on it.

**Comment by dns2utf8** [comment]
I moved the in-tree part to the README of the manual process, since it was just for convenience.

I was looking for the structure of the build bot, can I run one locally?

**Comment by steveklabnik** [comment]
@alexcrichton ping re last comment here, can @dns2utf8 do something to test this out?

**Comment by alexcrichton** [comment]
We've got a [number of docker images](https://github.com/rust-lang/rust-buildbot/tree/master/slaves) which our bots are running, so running the tests inside those containers would perhaps be enough to test out?

**Comment by dns2utf8** [comment]
I was able to start a container and access it with this command:
```bash
docker run --rm -it --entrypoint=/bin/bash alexcrichton/rust-slave-linux:2016-08-04
```
then inside clone my branch like the script:
```bash
git clone https://github.com/dns2utf8/rust.git --branch doc_grammar
```
I can then build my branch:
```bash
cd rust
./configure && make -j$(nproc) docs
```
but I am not sure this is the same as you do. 
Am I doing it wrong? @jonathandturner

**Comment by alexcrichton** [comment]
@dns2utf8 yeah that's basically the same as what we do, this'd just be testing for presence of the relevant tools.

Is this running by default on `make check` though? I thought you had a run a custom check target?

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with a rebase!

**Comment by dns2utf8** [comment]
Thanks for the reminder, I rebased it.

**Pull request closed without merging**